# Project Foundations for Data Science: FoodHub Data Analysis

**Marks: 60**

### Context

The number of restaurants in New York is increasing day by day. Lots of students and busy professionals rely on those restaurants due to their hectic lifestyles. Online food delivery service is a great option for them. It provides them with good food from their favorite restaurants. A food aggregator company FoodHub offers access to multiple restaurants through a single smartphone app.

The app allows the restaurants to receive a direct online order from a customer. The app assigns a delivery person from the company to pick up the order after it is confirmed by the restaurant. The delivery person then uses the map to reach the restaurant and waits for the food package. Once the food package is handed over to the delivery person, he/she confirms the pick-up in the app and travels to the customer's location to deliver the food. The delivery person confirms the drop-off in the app after delivering the food package to the customer. The customer can rate the order in the app. The food aggregator earns money by collecting a fixed margin of the delivery order from the restaurants.

### Objective

The food aggregator company has stored the data of the different orders made by the registered customers in their online portal. They want to analyze the data to get a fair idea about the demand of different restaurants which will help them in enhancing their customer experience. Suppose you are hired as a Data Scientist in this company and the Data Science team has shared some of the key questions that need to be answered. Perform the data analysis to find answers to these questions that will help the company to improve the business. 

### Data Description

The data contains the different data related to a food order. The detailed data dictionary is given below.

### Data Dictionary

* order_id: Unique ID of the order
* customer_id: ID of the customer who ordered the food
* restaurant_name: Name of the restaurant
* cuisine_type: Cuisine ordered by the customer
* cost: Cost of the order
* day_of_the_week: Indicates whether the order is placed on a weekday or weekend (The weekday is from Monday to Friday and the weekend is Saturday and Sunday)
* rating: Rating given by the customer out of 5
* food_preparation_time: Time (in minutes) taken by the restaurant to prepare the food. This is calculated by taking the difference between the timestamps of the restaurant's order confirmation and the delivery person's pick-up confirmation.
* delivery_time: Time (in minutes) taken by the delivery person to deliver the food package. This is calculated by taking the difference between the timestamps of the delivery person's pick-up confirmation and drop-off information

### Let us start by importing the required libraries

In [1]:
# import libraries for data manipulation
import numpy as np
import pandas as pd

# import libraries for data visualization
import matplotlib.pyplot as plt
import seaborn as sns

### Understanding the structure of the data

In [4]:
# read the data
# (This try/except makes the notebook portable across different working directories)
import os

possible_paths = ['foodhub_order.csv', '/mnt/data/foodhub_order.csv']
for p in possible_paths:
    if os.path.exists(p):
        df = pd.read_csv(p)
        break

df.head()


,order_id,customer_id,restaurant_name,cuisine_type,cost_of_the_order,day_of_the_week,rating,food_preparation_time,delivery_time
0,1477147,337525,Hangawi,Korean,30.75,Weekend,Not given,25,20
1,1477685,358141,Blue Ribbon Sushi Izakaya,Japanese,12.08,Weekend,Not given,25,23
2,1477070,66393,Cafe Habana,Mexican,12.23,Weekday,5,23,28
3,1477334,106968,Blue Ribbon Fried Chicken,American,29.20,Weekend,3,25,15
4,1478249,76942,Dirty Bird to Go,American,11.59,Weekday,4,25,24


#### Observations:
- The DataFrame matches the data dictionary and contains order-level information (one row per order).
- The dataset includes identifiers (order and customer), restaurant and cuisine information, order cost, order day type, customer rating, and operational time metrics (preparation and delivery times).

### **Question 1:** How many rows and columns are present in the data? [0.5 mark]

In [10]:
print('Shape (rows, columns):', df.shape)

Shape: (1898, 9)


#### Observations:
- The dataset contains **1,898 rows** and **9 columns**.
- Each row represents a unique food order placed by a customer through the FoodHub platform.

### **Question 2:** What are the datatypes of the different columns in the dataset? (The info() function can be used) [0.5 mark]

In [14]:
# Use info() to print a concise summary of the DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1898 entries, 0 to 1897
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   order_id               1898 non-null   int64  
 1   customer_id            1898 non-null   int64  
 2   restaurant_name        1898 non-null   object 
 3   cuisine_type           1898 non-null   object 
 4   cost_of_the_order      1898 non-null   float64
 5   day_of_the_week        1898 non-null   object 
 6   rating                 1898 non-null   object 
 7   food_preparation_time  1898 non-null   int64  
 8   delivery_time          1898 non-null   int64  
dtypes: float64(1), int64(4), object(4)
memory usage: 133.6+ KB


#### Observations:
- `order_id` and `customer_id` are integer identifier variables.
- `restaurant_name`, `cuisine_type`, `day_of_the_week`, and `rating` are categorical variables.
- `cost_of_the_order` is a continuous numerical variable.
- `food_preparation_time` and `delivery_time` are numerical time variables (in minutes).
- The `rating` column contains the value **"Not given"**, so it should be converted to numeric (with "Not given" treated as missing) for quantitative analysis.

### **Question 3:** Are there any missing values in the data? If yes, treat them using an appropriate method. [1 mark]

In [16]:
# Missing values per column
df.isna().sum().sort_values(ascending=False)


# Duplicate rows
print('Duplicate rows:', df.duplicated().sum())


order_id                 0
customer_id              0
restaurant_name          0
cuisine_type             0
cost_of_the_order        0
day_of_the_week          0
rating                   0
food_preparation_time    0
delivery_time            0
dtype: int64

#### Observations:
- There are **no missing values** in any column.
- There are also **no duplicate rows** in the dataset.
- Since missing values are not present, no imputation is required; however, `rating` requires cleaning because it includes a non-numeric category ("Not given").

### **Question 4:** Check the statistical summary of the data. What is the minimum, average, and maximum time it takes for food to be prepared once an order is placed? [2 marks]

In [ ]:
# Statistical summary for food preparation time
prep_min = df['food_preparation_time'].min()
prep_mean = df['food_preparation_time'].mean()
prep_max = df['food_preparation_time'].max()

print('Food preparation time (minutes) -> min:', prep_min, '| mean:', round(prep_mean, 2), '| max:', prep_max)


#### Observations:
- Food preparation time ranges from **20 minutes** to **35 minutes**.
- The average food preparation time is approximately **27.37 minutes**.
- Preparation time shows relatively low variability (narrow range), suggesting restaurant preparation times are fairly consistent overall.

### **Question 5:** How many orders are not rated? [1 mark]

In [ ]:
# Number of orders not rated
not_rated = (df['rating'] == 'Not given').sum()
total_orders = df.shape[0]
print('Orders not rated:', not_rated)
print('Percentage not rated:', round(not_rated/total_orders*100, 2), '%')


#### Observations:
- **736 orders** are not rated by customers.
- This represents approximately **38.78%** of all orders.
- Ratings analysis should therefore be performed on the subset of orders where ratings are available.

### Exploratory Data Analysis (EDA)

### Univariate Analysis

### **Question 6:** Explore all the variables and provide observations on their distributions. (Generally, histograms, boxplots, countplots, etc. are used for univariate exploration.) [9 marks]

In [ ]:
# Univariate Analysis

# Convert rating to numeric (Not given -> NaN) for numerical analysis
df['rating_num'] = pd.to_numeric(df['rating'], errors='coerce')

# Create total time from order confirmation to delivery drop-off
df['total_time'] = df['food_preparation_time'] + df['delivery_time']

# ---- Categorical variables ----
plt.figure(figsize=(6,4))
sns.countplot(data=df, x='day_of_the_week')
plt.title('Orders by Day Type')
plt.show()

plt.figure(figsize=(8,4))
df['cuisine_type'].value_counts().head(10).sort_values().plot(kind='barh')
plt.title('Top 10 Cuisines by Number of Orders')
plt.xlabel('Number of Orders')
plt.show()

plt.figure(figsize=(8,4))
df['restaurant_name'].value_counts().head(10).sort_values().plot(kind='barh')
plt.title('Top 10 Restaurants by Number of Orders')
plt.xlabel('Number of Orders')
plt.show()

plt.figure(figsize=(6,4))
sns.countplot(x=np.where(df['rating']=='Not given','Not given','Given'), data=df)
plt.title('Rating Availability')
plt.xlabel('Rating status')
plt.show()

# ---- Numerical variables ----
num_cols = ['cost_of_the_order','food_preparation_time','delivery_time','total_time','rating_num']

for col in num_cols:
    plt.figure(figsize=(6,4))
    sns.histplot(df[col].dropna(), kde=True, bins=30)
    plt.title(f'Distribution of {col}')
    plt.show()

    plt.figure(figsize=(6,2.5))
    sns.boxplot(x=df[col])
    plt.title(f'Boxplot of {col}')
    plt.show()


### **Question 7**: Which are the top 5 restaurants in terms of the number of orders received? [1 mark]

In [ ]:
# Top 5 restaurants by number of orders
top5_restaurants = df['restaurant_name'].value_counts().head(5)
top5_restaurants


#### Observations:
- The top 5 restaurants account for a substantial share of platform demand.
- The top 5 restaurants by number of orders are:
  - **Shake Shack** (219 orders)
  - **The Meatball Shop** (132 orders)
  - **Blue Ribbon Sushi** (119 orders)
  - **Blue Ribbon Fried Chicken** (96 orders)
  - **Parm** (68 orders)
- These restaurants are key partners for FoodHub and should be prioritized for operational support and promotional visibility.

### **Question 8**: Which is the most popular cuisine on weekends? [1 mark]

In [ ]:
# Most popular cuisine on weekends
weekend_orders = df[df['day_of_the_week'] == 'Weekend']
weekend_cuisine_counts = weekend_orders['cuisine_type'].value_counts()
weekend_cuisine_counts.head(10)


#### Observations:
- The most popular cuisine on weekends is **American**.
- This suggests weekend demand is strongly driven by this cuisine type, which can be leveraged for weekend-specific promotions and restaurant onboarding strategies.

### **Question 9**: What percentage of the orders cost more than 20 dollars? [2 marks]

In [ ]:
# Percentage of orders costing more than $20
pct_more_20 = (df['cost_of_the_order'] > 20).mean() * 100
round(pct_more_20, 2)


#### Observations:
- Approximately **29.24%** of orders cost more than **$20**.
- These higher-value orders likely contribute disproportionately to FoodHub revenue; ensuring fast, reliable delivery for these orders can protect customer satisfaction and revenue.

### **Question 10**: What is the mean order delivery time? [1 mark]

In [ ]:
# Mean order delivery time (in minutes)
mean_delivery_time = df['delivery_time'].mean()
round(mean_delivery_time, 2)


#### Observations:
- The mean delivery time is approximately **24.16 minutes**.
- Delivery time is a key operational metric that can strongly influence customer satisfaction and ratings.

### **Question 11:** The company has decided to give 20% discount vouchers to the top 3 most frequent customers. Find the IDs of these customers and the number of orders they placed. [1 mark]

In [ ]:
# Top 3 most frequent customers
top3_customers = df['customer_id'].value_counts().head(3).reset_index()
top3_customers.columns = ['customer_id','number_of_orders']
top3_customers


#### Observations:
- The top 3 most frequent customers are:
  - **Customer 52832** with **13 orders**
  - **Customer 47440** with **10 orders**
  - **Customer 83287** with **9 orders**
- These customers are strong candidates for loyalty benefits (e.g., discount vouchers) to improve retention and lifetime value.

### Multivariate Analysis

### **Question 12**: Perform a multivariate analysis to explore relationships between the important variables in the dataset. (It is a good idea to explore relations between numerical variables as well as relations between numerical and categorical variables) [10 marks]


In [ ]:
# Multivariate Analysis

# 1) Correlation between numerical variables (excluding IDs)
num_for_corr = df[['cost_of_the_order','food_preparation_time','delivery_time','total_time','rating_num']].copy()
corr = num_for_corr.corr()

plt.figure(figsize=(7,5))
sns.heatmap(corr, annot=True, fmt='.2f', cmap='Blues')
plt.title('Correlation Heatmap (Numerical Variables)')
plt.show()

# 2) Delivery time by weekday vs weekend
plt.figure(figsize=(6,4))
sns.boxplot(data=df, x='day_of_the_week', y='delivery_time')
plt.title('Delivery Time by Day Type')
plt.show()

# 3) Total time vs rating
rated_df = df.dropna(subset=['rating_num'])
plt.figure(figsize=(6,4))
sns.scatterplot(data=rated_df, x='total_time', y='rating_num', alpha=0.5)
plt.title('Rating vs Total Time')
plt.xlabel('Total time (minutes)')
plt.ylabel('Rating (1–5)')
plt.show()

# 4) Cost distribution by cuisine (top 6 cuisines)
top_cuisines = df['cuisine_type'].value_counts().head(6).index
plt.figure(figsize=(10,4))
sns.boxplot(data=df[df['cuisine_type'].isin(top_cuisines)], x='cuisine_type', y='cost_of_the_order')
plt.title('Order Cost by Cuisine (Top 6)')
plt.xticks(rotation=30, ha='right')
plt.show()

# 5) Total time by day type
plt.figure(figsize=(6,4))
sns.boxplot(data=df, x='day_of_the_week', y='total_time')
plt.title('Total Time (Prep + Delivery) by Day Type')
plt.show()


### **Question 13:** The company wants to provide a promotional offer in the advertisement of the restaurants. The condition to get the offer is that the restaurants must have a rating count of more than 50 and the average rating should be greater than 4. Find the restaurants fulfilling the criteria to get the promotional offer. [3 marks]

In [ ]:
# Restaurants eligible for promotional offer:
# rating_count > 50 and average_rating > 4

rating_summary = rated_df.groupby('restaurant_name').agg(
    rating_count=('rating_num','count'),
    avg_rating=('rating_num','mean')
).reset_index()

eligible = rating_summary[(rating_summary['rating_count'] > 50) & (rating_summary['avg_rating'] > 4)]     .sort_values('avg_rating', ascending=False)

eligible


#### Observations:
- After excluding unrated orders, **4 restaurants** meet the criteria (**rating count > 50** and **average rating > 4**) and are eligible for promotional offers.
- These restaurants combine **high volume of feedback** with **strong customer satisfaction**, making them ideal for advertising campaigns and in-app highlighting.

### **Question 14:** The company charges the restaurant 25% on the orders having cost greater than 20 dollars and 15% on the orders having cost greater than 5 dollars. Find the net revenue generated by the company across all orders. [3 marks]

In [ ]:
# Net revenue generated by the company across all orders
# Rule:
# - 25% commission on orders with cost > $20
# - 15% commission on orders with cost > $5 (and <= $20)
# - 0% otherwise

cost = df['cost_of_the_order']

commission_rate = np.select(
    [cost > 20, cost > 5],
    [0.25, 0.15],
    default=0.0
)

df['commission'] = cost * commission_rate
net_revenue = df['commission'].sum()

round(net_revenue, 2)


#### Observations:
- Using the commission rules provided, the estimated **net revenue** (commission) generated across all orders is **$6,166.30**.
- Revenue is higher for orders above $20 due to the higher commission rate (25%), so high-value orders are strategically important for the platform.

### **Question 15:** The company wants to analyze the total time required to deliver the food. What percentage of orders take more than 60 minutes to get delivered from the time the order is placed? (The food has to be prepared and then delivered.) [2 marks]

In [ ]:
# Percentage of orders taking more than 60 minutes total (prep + delivery)
pct_over_60 = (df['total_time'] > 60).mean() * 100
round(pct_over_60, 2)


#### Observations:
- Approximately **10.54%** of orders take more than **60 minutes** in total (preparation + delivery).
- These slow orders are likely to negatively impact customer satisfaction; identifying root causes (restaurant prep delays vs delivery delays) can help reduce this percentage.

### **Question 16:** The company wants to analyze the delivery time of the orders on weekdays and weekends. How does the mean delivery time vary during weekdays and weekends? [2 marks]

In [ ]:
# Mean delivery time on weekdays vs weekends
mean_delivery_by_day = df.groupby('day_of_the_week')['delivery_time'].mean().round(2)
mean_delivery_by_day


#### Observations:
- Mean delivery time differs by day type:
  - **Weekday:** 28.34 minutes  
  - **Weekend:** 22.47 minutes
- In this dataset, delivery time is **higher on weekdays**, which may reflect weekday traffic patterns and delivery density during workdays.

### Conclusion and Recommendations

### **Question 17:** What are your conclusions from the analysis? What recommendations would you like to share to help improve the business? (You can use cuisine type and feedback ratings to drive your business recommendations.) [6 marks]

### Conclusions:
- The FoodHub dataset contains 1,898 orders across multiple restaurants and cuisine types. Demand is concentrated among a small set of restaurants, indicating that a few partners drive a large portion of the platform’s activity.
- A significant share of orders (about two-fifths) are not rated, so rating-based insights must be derived from the subset of rated orders.
- Total fulfillment time (preparation + delivery) is a key driver of the customer experience. Orders taking more than 60 minutes represent a meaningful minority and should be targeted for operational improvement.
- High-value orders (cost > $20) form an important segment because they attract a higher commission rate and likely contribute disproportionately to net revenue.
- Weekday and weekend patterns differ, suggesting opportunities for demand-aware staffing and delivery resource allocation.

### Recommendations:
- **Prioritize high-demand restaurants:** Provide operational support (e.g., faster pick-up coordination) and feature them strategically in-app because they generate a large share of orders.
- **Improve speed to protect ratings:** Track and reduce total time (prep + delivery). Investigate slow orders to determine whether delays are driven by restaurant preparation or delivery logistics.
- **Weekend and weekday staffing optimization:** Adjust delivery partner availability by day type based on observed delivery-time differences and order volumes.
- **Promote high-performing restaurants:** Restaurants with high rating counts and strong average ratings should be highlighted in advertisements and app rankings to improve conversion and trust.
- **Protect high-value orders:** Since orders > $20 generate higher commission, ensure reliable delivery performance for these orders (e.g., priority dispatching) to protect revenue and satisfaction.

---